### ISSUE TITLE
Tracking issue for panics in mem::uninitialized/zeroed

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#66151]
With https://github.com/rust-lang/rust/pull/66059, `mem::uninitialized` and `mem::zeroed` dynamically detect some misuses (such as `mem::zeroed::<&T>()` or `mem::uninitialized::<bool>()`) and panic instead of causing UB. Also see [this summary for the original FCP](https://github.com/rust-lang/rust/pull/66059#issuecomment-586734747). But the check is conservative for now to reduce breakage. This is to track strengthening the check.

* [x] We should recursively check the fields of structs and tuples. This is happening in https://github.com/rust-lang/rust/pull/71274.
* [ ] We should check inside arrays, too. We currently do not because smallvec < 0.6.13 didn't use `MaybeUninit` and thus triggers this check (if the type inside the array has invalid bit patterns). Other widely-used crates that trigger the panic as of [Dec 2021](https://github.com/rust-lang/rust/pull/87041#issuecomment-997587451):
    * hyper ([fixed](https://github.com/hyperium/hyper/pull/2545) since 0.14.12)
    * crossbeam-queue 0.2.1, crossbeam-channel 0.3.9, crossbeam 0.2.12 (these are ancient; current versions are fine AFAIK)
* [ ] We should check `Variant::Multiple` enums, which currently we do not just to stay conservative.

**Comment by RalfJung** [comment]
git2-rs [needed some adjustment](https://github.com/rust-lang/git2-rs/commit/d70b64599a080ea91f6052cb234a4a485f5dcdff) to work even with the minimal checks that have already landed, which is good enough for now but even those adjustments will nut suffice once we properly check enums. The reason for this is that [ctest uses `mem::uninitialized()`](https://github.com/gnzlbg/ctest/issues/90) to test that the bindings are declared correctly, and an uninitialized `enum` (with more than one variant in its layout) is UB. It has to be UB because the point of the validity invariant of an `enum` is to make sure the discriminant is always valid, and for uninitialized memory the discriminant test would raise UB.

Cc @gnzlbg @alexcrichton

**Comment by valarauca** [comment]
Will the unstable methods associated with `MaybeUninit` be stabilized along with or as part of this change?

I agree this is a positive change for the language, but with a large amount of the `MaybeUninit` API unstable/nightly only this change feels premature. There's a lot of patterns which simply are not allowed in non-nightly `MaybeUninit`, and having it suggested as an upgrade path from `mem::uninitialized/zeroed` feels a bit like a bad joke.

**Comment by RalfJung** [comment]
@valarauca `MaybeUninit` has been stable for quite a while. The remaining unstable methods are just convenience, they do not limit functionality. For example, instead of `get_mut` (just renamed to `assume_init_mut`) you can do `&mut *x.as_mut_ptr()`. The safety rules are the same as for `assume_init_mut`: the data must be already fully initialized before you create a mutable reference.

Which patterns specifically are you talking about?

**Comment by Aaron1011** [comment]
In https://github.com/rust-lang/rust/issues/77970#issuecomment-713204477, the panic in `mem::uninitialized` ended up causing a segfault, since unwinding caused some unsafe drop code to run in an invalid state.

I think it might make sense to force an abort (with a message), rather than use whatever panic strategy is configured. There's some precedent for this - both `Rc` and `Arc` will abort in the face of a pathological number of `clone()` calls (which can actually be triggered from safe code). Since this panic occurs when unsafe code has already triggered undefined behavior, catching the panic or unwinding will probably just make things worse.

**Comment by scottmcm** [comment]
From https://github.com/rust-lang/rust/pull/66059#issuecomment-586734747:
> It is not entirely clear whether, when a violation of validity is detected, we should panic or abort. For now, we panic, which has the advantage of printing a backtrace where possible. However, unsafe could could be surprised by the panic and not expect unwinding, and that has lead to segfaults on crater. The issue with aborting, however, is that we have no good way to print a backtrace.

**Comment by Aaron1011** [comment]
There's ongoing work to make `Backtrace` available in `core`: https://github.com/rust-lang/rust/pull/72981#issuecomment-707405697

Once that becomes available on nightly, we could use it internally to attempt to print a backtrace, and then abort.

**Comment by RalfJung** [comment]
> Once that becomes available on nightly, we could use it internally to attempt to print a backtrace, and then abort.

Yes, that would certainly be worth trying! I ran into such unexpected-unwind segfaults before when analyzing crater runs for this change and they are indeed quite hard to debug.

**Comment by nico-abram** [comment]
It's been about a year since this issue was created. Has enough time passed that PRs adding the check for arrays and enums would be merged?

**Comment by RalfJung** [comment]
We *just* landed this check for structs and tuples on the stable branch with Rust 1.48. The fallout from that is still rippling through the ecosystem, e.g. [here](https://github.com/zslayton/lifeguard/issues/25#issuecomment-730316210). So I feel we should wait a bit before retching this up another level.

@nico-abram would you be interested in preparing such a change for arrays and seeing it through (in particular analyzing the crater report)? The code change for arrays is trivial, and I can provide mentoring, but last time analyzing the crater fallout took forever so I won't do this again anytime soon.^^

**Comment by nico-abram** [comment]
Yes! I can, assuming there is no set timeline required

If I understand correctly, the change would be replacing [this comment](https://github.com/rust-lang/rust/blob/38030ffb4e735b26260848b744c0910a5641e1db/compiler/rustc_target/src/abi/mod.rs#L1166) with a similar check to the `FieldsShape::Arbitrary` one but hardcoded for index 0, and an or check for `FieldsShape::Array { count, .. }` count == 0 (Like there is for `Abi::Vector`). Is that correct?

The change would just be a straight up PR against this repository, and someone authorized would schedule a crater run?

Finally, does analyzing the crater report mean looking into the errors and checking why they happen, analyzing the ecosystem-wide impact of breakage, and contacting maintainers with bug reports?

**Comment by RalfJung** [comment]
> Yes! I can, assuming there is no set timeline required

No timeline required. :) Thanks!

> If I understand correctly, the change would be replacing this comment with a similar check to the FieldsShape::Arbitrary one but hardcoded for index 0, and an or check for FieldsShape::Array { count, .. } count == 0 (Like there is for Abi::Vector). Is that correct?

Yes, that sounds right: first determine if the array has at least one element, then check the layout of the element type.

The [`Array` variant](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/enum.FieldsShape.html#variant.Array) has a `count` field for the element count, as you noted.

> The change would just be a straight up PR against this repository, and someone authorized would schedule a crater run?

Exactly. The PR should also extend the test "src/test/ui/intrinsics/panic-uninitialized-zeroed.rs" to make sure the check behaves as expected.

> Finally, does analyzing the crater report mean looking into the errors and checking why they happen, analyzing the ecosystem-wide impact of breakage, and contacting maintainers with bug reports?

Basically, yes. You can see my analysis for the last change [here](https://github.com/rust-lang/rust/pull/71274#issuecomment-683937298).  As you can see from the bingbacks above that comment, there were quite a few bugreports to be made for this one.

I believe that the fallout will be smaller this time, since I think that we covered most of the "heavy hitters" last time and they wouldn't show up as regressions any more now (since they are already broken on stable). However, there might well be unexpected things showing up (like vast numbers of people still using outdated smallvec, which is known to be broken).

**Comment by RalfJung** [comment]
Here's a good testcase for an enum that we should detect is UB for `mem::uninitialized`: https://github.com/rust-lang/rust/issues/79282#issuecomment-731845834.

**Comment by Aaron1011** [comment]
We might want to consider enabling the array checks only for calls made in 2021 crates. This could be accomplished by ensuring that `mem::uninitialized` and `mem::zeroed` are always instantiated on the current crate when called with an invalid type (e.g. generics sharing will never apply), and then checking the crate edition during codegen of the intrinsic.

This should hopefully cause much less breakage than revealed in https://github.com/rust-lang/rust/pull/87041, and help prevent any new unsound code from being written using these APIs.

**Comment by 5225225** [comment]
Could we start showing a future compat warning when people leave something uninit? (which I believe shows to the user even if it's in a dependency?) That might help cut down on the people who continue to use outdated dependencies (notably, tokio-fs shows up in the above crater run, and gets around 4k downloads a day despite not being updated in years (since it was moved to `tokio`), which is concerning)

I'm not sure what
> We should check Variant::Multiple enums, which currently we do not just to stay conservative.
means, since Result<u8, u8> panics. Does that mean repr(u8) enums that have all 256 variants filled?

**Comment by Aaron1011** [comment]
Unfortunately, there are legitimate reasons to have a call to `std::mem::{uninitialized, zeroed}` with an invalid type. You could have a macro or generic function that contains such a call, but gated behind a check:

```rust
macro_rules! maybe_zero {
    ($t:type) => {
        if MyTrait::can_zero_init::<$t>() {
            Some(std::mem::zeroed::<$T>())
        } else {
            None
        }
    }
}
```

We could try to warn on code that appears to come from the same expansion context (to catch something like an explicitly written `std::mem::uninitialized::<bool>()`, but there could still be false positives due to proc-macros reassgning spans.

**Comment by 5225225** [comment]
Hm, good point.

Though we *do* already have a warning for literally writing `std::mem::uninitialized::<bool>()` (which *does* trigger with that macro if it's in the same crate as you), so that *could* be expanded to warn on arrays, I suppose? Depends how many false positives we'd expect from that.

Though I'm not sure how much that would help, since basically *all* of the panics come from old dependencies.

**Comment by RalfJung** [comment]
Yeah, extending the lint to arrays should be rather uncontroversial at this point. That would be an `Array` case in [this `match`](https://github.com/rust-lang/rust/blob/d54195db22421c51fd14560aba3bbf9b79a52677/compiler/rustc_lint/src/builtin.rs#L2535).

> Does that mean repr(u8) enums that have all 256 variants filled?

Indeed such an enum can *not* be left uninitialized.

**Comment by RalfJung** [comment]
> Yeah, extending the lint to arrays should be rather uncontroversial at this point. 

I implemented that in https://github.com/rust-lang/rust/pull/95373.

**Comment by RalfJung** [comment]
> We might want to consider enabling the array checks only for calls made in 2021 crates.

After having another look at the report in https://github.com/rust-lang/rust/pull/87041, I tend to agree -- I am not sure if old hyper and crossbeam will ever go sufficiently out of use that we can break them entirely. Hopefully when people bump the edition of their crate they run the test suite in a recent compiler as well (or CI does it for them) and that way they learn about the problem. However, there is no precedent for codegen changes on an edition boundary and this could be rather surprising, so I wonder what the lang team thinks about this... let's [ask them on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Making.20mem.3A.3Auninitialized.20detect.20more.20UB.20only.20in.20new.20edition).

**Comment by Aaron1011** [comment]
The closest precedent for codegen changes would be min captures for closures - IIRC, there's a lint that warns you if the drop order will get changed when you upgrade to edition 2021.

**Comment by RalfJung** [comment]
Regarding the problem with old hyper and crossbeam, @oli-obk suggested
> We could backport patches to these old crates... I'm sure we can get the maintainers to do that if we do the impl work

The trouble is, the old version is in tons of lockfiles, which crater respects (and generally for good reasons, since cargo will also respect them when people build those crates). I think a lot of these are essentially unmaintained, so I doubt we'll ever get all those lockfiles changed. I assume we are pretty much stuck with supporting old hyper and crossbream for forever (if optimizations break that code one day, I guess we'll see how we handle that).

We *could* do something terrible and hard-code those two crates as exceptions in these checks...
Or we could do something edition-based.

However, as was pointed out on Zulip, one issue here is that the intrinsic is actually called inside a std function, so the name or edition of the crate calling `std::mem::uninitialized` is not even known. I am not sure if there is a way around that.

**Comment by saethlin** [comment]
Do we check enums yet? If special-casing crates is even close to being on the table, we should at least try checking enums. I skimmed through crates that call `mem::uninitialized` and I'm pretty sure [that would at least "break" stdweb](https://github.com/koute/stdweb/blob/9b418d98df6fafaa4d4b87b04c304d0220292055/src/webcore/serialization.rs).

**Comment by RalfJung** [comment]
If the enum has `Scalar` or `ScalarPair` layout, we check it. (In particular this means we check fieldless enums.)

If the enum has more complicated layout, we don't. I don't think a crater run for that has been tried. It's non-trivial to determine what the valid layouts of a complicated enum are due to niche optimizations. However it should never be valid to have uninit in a multi-variant type (niche optimization or not) so that might be a start.

stdweb might manage to avoid that issue by having `repr(u8)` on the enum, which might disable the niche (I am not sure). [This](https://github.com/koute/stdweb/blob/9b418d98df6fafaa4d4b87b04c304d0220292055/src/webcore/serialization.rs#L415) is still definitely wrong though, and should be fairly easy to replace with `MaybeUninit`.

**Comment by saethlin** [comment]
Also FWIW, it seems that the outcome of crashing or deprecating `mem::uninitialized` has been mixed, because [some](https://miri.saethlin.dev/ub?crate=generator&version=0.7.0) [crates](https://miri.saethlin.dev/ub?crate=blst&version=0.3.7) [just](https://miri.saethlin.dev/ub?crate=direction&version=0.18.1) [replaced](https://miri.saethlin.dev/ub?crate=ntex-bytes&version=0.1.14) [it](https://miri.saethlin.dev/ub?crate=yazi&version=0.1.4) [with](https://miri.saethlin.dev/ub?crate=parity-db&version=0.3.11) [the](https://miri.saethlin.dev/ub?crate=ya-client-model&version=0.4.0) [wrong](https://miri.saethlin.dev/ub?crate=unqlite&version=1.5.0) [thing](https://miri.saethlin.dev/ub?crate=zstd-seekable&version=0.1.7) which I'm told we can't detect statically or at runtime without Miri. I feel like we're doing the righteous thing here, but if it's causing people to write harder-to-detect UB... idk.

**Comment by RalfJung** [comment]
Easier-to-detect UB doesn't really help if you choose not to detect it. :shrug: 

FWIW, our lint against constructing invalid values also fires on `MaybeUninit::uninit().assume_init()`. But so far we chose not to lint when it's just constructing integers. And of course if things are too generic we cannot lint either.

**Comment by RalfJung** [comment]
@saethlin btw I think it would be a great idea to go over the instances of `MaybeUninit::uninit().assume_init()` you found, see if the lint detects them, and if not -- extend the lint to detect them.

OTOH, if people do [things like that](https://github.com/ntex-rs/ntex/blob/fb6c16fcb2c9f9f48abfc264d63b4641251f30de/ntex-bytes/src/bytes.rs#L2965)
```rust
        #[allow(invalid_value, clippy::uninit_assumed_init)]
        let mut inner: Inner = mem::MaybeUninit::uninit().assume_init();
```
then we can just shrug, I think -- we did what we could to help them, if they *want* to shoot themselves into the foot, we cannot stop them.

**Comment by RalfJung** [comment]
FWIW there is already an unrelated [security advisory](https://github.com/rustsec/advisory-db/blob/main/crates/hyper/RUSTSEC-2021-0079.md) against hyper that is fixed with 0.14.10, so almost all the crates that lead to panics due to hyper are already showing warnings if anyone were to ever run `cargo audit` on them. (The fix for uninit memory landed with 0.14.12, but most crates that I found failing used versions older than 0.14.10.)

For crossbeam, I filed https://github.com/rustsec/advisory-db/pull/1231.

**Comment by 5225225** [comment]
I was looking into this again, and wrote this pipeline to show only unique failures in dependencies: `rg '(attempted to .* which is invalid).*(registry/src/.*?):' -I -o -r '$2' | sort | uniq`

Running it on the regressions from the historical crater run (and only listing semver compatible versions once):

`array-init 0.0.4`, `arrayvec 0.3, 0.4`, `coco 0.1`, `crossbeam 0.2`, `crossbeam-channel 0.2, 0.3, 0.4`, `crossbeam-deque 0.7`, `crossbeam-epoch 0.8`, `crossbeam-queue 0.1, 0.2`, `elastic-array 0.7`, `elastic-array-plus 0.9, 0.10`, `hyper 0.11, 0.12, 0.13, 0.14`, `smallvec 0.4, 0.6`, `tokio-sync 0.1` (Some of these may have fixed versions for that semver version, I know hyper does)

That doesn't find direct usages, but if all these root causes are fixed, I expect there to not be many crates that are directly doing a bad mem::uninit. And that's not *too* many backports. (array-init i know has been fixed by now, and 0.0.4 has only one user that ran into the panic on the crater run)

Cursed idea: the panic runtime knows the calling crate, right? So

1. do backports to all the majorly used crates there, or at least as many as possible
2. modify the panic to somehow know the crate name (and ideally version, but i don't think rustc knows about crate versions, that's a cargo thing) and go "hey, you can try running `cargo update -p hyper` to get an updated version". This might require branching in the assertion based on the caller_location

This will still break code that has lockfiles set but it will make it clear what the crate developer needs to do to fix it.

**Comment by 5225225** [comment]
Also, that crater run didn't consider uninit `u8`s to be UB, which I know `http` did do (now fixed).

So if we want the panics to match what we call UB, I expect there to be 2 or 3 more phases of making them stricter

1. Just looking inside arrays
2. Uninit types that have no validity invariant (other than not being uninit) are UB too
3. (Maybe) that check that was mentioned in the comments about "Just make an uninit/zeroed type and ask the const eval if it's okay", I expect this to find maybe a few edge cases but nothing major.

Somewhere in here I think should be just straight up removing mem::uninitialized, or making it a hard error/forbid by default with future compat warning to call with a type that will panic. Maybe do the same to mem::zeroed, forbid by default with a FCW.

That is a breaking change (but so is adding these panics in the first place, not that I think we *shouldn't*), but I feel it's probably a nicer behaviour than letting it panic at runtime.

It does come with the cost that, yes, you can't use `if T::can_be_uninit() { mem::uninitialized() }`, but that would only be an issue for library crates with a MSRV lower than MaybeUninit who's used by newer crates, since I suppose the recommended replacement would be `MaybeUninit::uninit().assume_init()`

Treating mem::uninitialized differently from MaybeUninit::uninit().assume_init() feels somewhat weird, but treating all uses of mem::uninit with suspicion feels justified since there's so few cases where it's *right*, since they're either making a ZST, or they're after unions were created but *before* MaybeUninit was created, and they made their own MaybeUninit.

That, or at the very least do it across an edition, edition 2024+ crates just can't call mem::uninit or mem::zeroed

**Comment by RalfJung** [comment]
> array-init 0.0.4, arrayvec 0.3, 0.4, coco 0.1, crossbeam 0.2, crossbeam-channel 0.2, 0.3, 0.4, crossbeam-deque 0.7, crossbeam-epoch 0.8, crossbeam-queue 0.1, 0.2, elastic-array 0.7, elastic-array-plus 0.9, 0.10, hyper 0.11, 0.12, 0.13, 0.14, smallvec 0.4, 0.6, tokio-sync 0.1 (Some of these may have fixed versions for that semver version, I know hyper does)

That is some nice data! Are the latest (potentially semver-incompatible) versions of all these crates fixed? Many of these already have Miri on CI; would it be worth adding it to the rest of them as well just to ensure this does not regress?

I doubt many of the lockfiles will ever get updated, so even if we had semver-compatible fixes for all these crates, crater would produce lots of failures. So, we'd need a policy change to get that landed.

> Somewhere in here I think should be just straight up removing mem::uninitialized

There was some support for removing it in the next edition. This just needs someone to work on adding that feature to rustc. :)  Probably it would take the form of some magic attribute that a standard library function can use to only be available in old editions.

**Comment by 5225225** [comment]
[array-init](https://github.com/Manishearth/array-init): latest 2.0.0, passes miri with its tests (uses MaybeUninit correctly, as far as I can see). That, and we're getting that functionality as part of the stdlib. Does not have miri in CI.

[arrayvec](https://github.com/bluss/arrayvec/): latest `0.7.2`, passes miri, looks to run miri in CI with default settings, which includes number validity now.

coco: deprecated, last update 4 years ago, github link is dead. Miri fails with a "creating a null pointer", `RUSTFLAGS="-Zstrict-init-checks" cargo test` segfaults, presumably it's not expecting a panic we introduced there.

crossbeam: (just going to write crossbeam for all of the crossbeam crates, they all look to be tested together) `cargo miri test` passes fine if you ignore leaks, they have miri pinned because of weak memory issues. strict-init-checks works fine.

[elastic-array](https://github.com/paritytech/elastic-array): latest 0.11.0, archived repo. Miri passes fine, as does strict-init-checks.

[elastic-array-plus](https://github.com/sorpaas/elastic-array): latest 0.10.0, not archived but also doesn't look very active (last commit 2018). Git version fails, making an uninit `[u8; 2]` in their tests.

hyper: uses miri in CI. strict-init-checks *does* find something

> thread 'client::connect::http::tests::local_address' panicked at 'attempted to leave type `*mut libc::unix::linux_like::ifaddrs` uninitialized, which is invalid', [...]/pnet_datalink-0.27.2/src/unix_interfaces.rs:38:45

this is testing from hyper's `master` branch. Running tests on [libpnet](https://github.com/libpnet/libpnet)'s `pnet_datalink` checked out to https://github.com/libpnet/libpnet/tree/v0.27.2 with -Zstrict-init-checks finds this panic, but `cargo miri test` does *not*. But it does say the main thread terminated without waiting for remaining threads. Strange. In any case, a newer release of libpnet has been made that seems to have fixed that. (and turns out it was me who fixed it, i remember working on this crate a while back `:)`), so hyper just needs to bump that dependency.

[smallvec](https://github.com/servo/rust-smallvec): passes miri fine, has miri in CI

tokio-sync: deprecated, moved into `tokio` proper. The crate that exists has tests that fail with strict init checks.

**Comment by Mark-Simulacrum** [comment]
Discussed briefly in lang backlog bonanza; we think that this no longer needs lang oversight. The remaining work here is gradually ratcheting up our checking for UB (and turning that into panics), but that doesn't need lang oversight.

**Comment by RalfJung** [comment]
> edition 2024+ crates just can't call mem::uninit

I have created an issue for that: https://github.com/rust-lang/rust/issues/98862

**Comment by 5225225** [comment]
Also did some looking to find direct test failures in the crater run, and came up with `rg '(attempted to .* which is invalid)'\'', src.*'` to just find panics caused in the same crate as the tests.

There's far fewer of these

https://crates.io/crates/keypad/0.1.4
https://crates.io/crates/par-array-init/0.0.5
https://crates.io/crates/hash-trie/0.4.0
https://crates.io/crates/hexagon/0.1.2
https://crates.io/crates/rudy/0.1.0

And two github ones

https://github.com/mvojacek/rust-expr-calc/tree/fe285145034d08d383d362c3f5f6c61aa72daf6c
https://github.com/nareshganduri/sudoku-solver-rs/tree/ebf9087b4385681872ff45b23a837486c10bc4a9

With the exception of hash-trie, all of these are at least 3+ years since the last release. (hash-trie is 1 year)

So I'm not too concerned with *direct* usage causing large amounts of panics. That, or the kind of crates that use mem::uninit unsoundly tend not to be on crates.io / not have tests.

**Comment by 5225225** [comment]
Hmm... is there a reason the invalid_value lint uses its own logic, instead of the may_permit_raw_init? Other than it giving better error messages, which could be added to may_permit_raw_init

What i'm thinking is

1. Make invalid_value use may_permit_raw_init on strict mode (without regressing the error messages)

2. Make invalid_value a FCW. This'll want to be done a while after, since we would have increased the strictness of the checks. Mention that this will panic at runtime / will in a future rustc.

3. Let the dust settle and have the maintained libraries/binaries notice the FCW and update

4. Go for another crater run, either for arrays or arrays + primitives panicking. If there are not too many issues / all the issues are for crates that Aren't getting updates, we'd presumably be good to add the panics.

I still feel a FCW on invalid_value is the best way forward, despite not being a perfect predictor of panics (because dead code / runtime checks). It's really the only way we have to get *everyone* to notice that one dependency that misuses mem::uninit *before* it panics at runtime.

And the crater run for arrays has *less* regressions than the one for aggregates did, so I don't feel this is too many, tbh.

**Comment by RalfJung** [comment]
> Add primitive checks to invalid_value.

You mean, making it warn on integers?
That makes sense. It would be a lang-team commitment to resolve https://github.com/rust-lang/unsafe-code-guidelines/issues/71 by saying these are indeed UB when uninit. So this would probably be a good first step. It should come with suitable doc updates; e.g. the `MaybeUninit` does currently say "the rules around uninitialized integers are not finalized yet".

Do you want to make such a PR? I am fully in support of this. I am not sure if the lang team is, but we *do* want `noundef` on all those integer argument types some day, so...

> Make invalid_value a FCW.

Note that `invalid_value` also fires on invalid uses of `MaybeUninit::uninit().assume_init()`. So I don't think this is the right plan.
Rather, we should probably start going down the road of https://github.com/rust-lang/rust/issues/98862, and make *any* use of `mem::uninitialized` be a FCW. If we can get consensus for that. Or, as an alternative, any definitely invalid use of `mem::uninitialized`.

**Comment by saethlin** [comment]
> we _do_ want `noundef` on all those integer argument types some day, so...

If a plan to do this is decided upon (which I would support, for what it's worth) I think we should plan ahead to take a stronger stance on defusing the UB that may emerge from this, even if that means we end up with new build or test failures in crater. I worry a lot about the incentive structure that crater is creating here: delivering surprise UB via new LLVM attributes won't show up in crater but a check that turns a security vulnerability into a test failure or a build break will.

With regard to checking for uninit integers, be aware that the patch which ported `itoa` from `mem::uninitialized` to `MaybeUninit` was released as version `1.0.0`, so for example anyone using the latest version of `csv` (currently `1.1.6`) would see this FCW. I think there are a fair few other crates which have released similar patches in semver-incompatible version bumps. I understand the MSRV reasoning for such behavior, but it's going to complicate things for us here because we cannot necessarily request crate authors release patches against unsupported versions which maintain the express or implied MSRV.

**Comment by 5225225** [comment]
> You mean, making it warn on integers?

Yeah, integers, floats, and raw pointers are the last things that we say are UB that we don't check.

Though `Option<usize>`, for example, doesn't seem to be warned about (to be uninit), although that panics at runtime. I'll look into the enum handling there too.

> Note that `invalid_value` also fires on invalid uses of `MaybeUninit::uninit().assume_init()`.

Ah, yeah, it also warns on transmutes which is cool. (though in theory we could do some magic data flow analysis stuff to notice that it couldn't possibly be init, and insert a panic instead, that's out of scope for *this* issue. maybe in a few years)

Though to be clear I do mean a FCW that is for all editions, as a "this is UB and we will make this panic soon, fix it". Since otherwise we shouldn't put noundef on primitives since we're not telling people about the UB in their dependency tree.

(So for the same reason I think it should only warn on UB cases, I don't care if someone's making an uninit ZST or union)

And it should also warn on UB uses of mem::zeroed for a similar reason, noundef doesn't apply, but validity invariants do, so we want to make that panic as well.

**Comment by RalfJung** [comment]
> (So for the same reason I think it should only warn on UB cases, I don't care if someone's making an uninit ZST or union)

OTOH there's a lot of code that uses `mem::uninitialized` on some generic type `T`. We'll never know for sure whether such code is UB. So in terms of future compatibility I'd rather get rid of `mem::uninitialized` entirely, either in some future edition or (if we can) on all editions. But anyway, that's another discussion -- that's https://github.com/rust-lang/rust/issues/98862.

**Comment by RalfJung** [comment]
> Though Option<usize>, for example, doesn't seem to be warned about (to be uninit), although that panics at runtime. I'll look into the enum handling there too.

Yeah that would be great. Basically `is_multi_variant` needs to use something like this to count the definitely existing variants
```rust
adt.variants().iter().filter(|v| v.fields.iter().all(|f| definitely_inhabited(f))).count()
```
and then `definitely_inhabited` needs to say (in particular) that `usize` is definitely inhabited.

**Comment by 5225225** [comment]
That can be future work after #98919 is merged, I assume?

I assume we don't know if it's inhabited for generic/extern type reasons, I suppose? Looks like we don't really know much about the type's layout at that point.

**Comment by RalfJung** [comment]
It's independent, so you can work on it in parallel if you don't want to wait for the decision process in https://github.com/rust-lang/rust/pull/98919.

> I assume we don't know if it's inhabited for generic/extern type reasons, I suppose? Looks like we don't really know much about the type's layout at that point.

Yeah, the safe fallback in `definitely_inhabited` is `false`. Currently it behaves as if that function *always* returned `false` but we can do better than that.

**Comment by RalfJung** [comment]
> And the crater run for arrays has less regressions than the one for aggregates did, so I don't feel this is too many, tbh.

For the aggregate one, I went over *every single crate* (not the GH projects), identified the root cause as well as I could, and made bug reports to notify the author of the regression. See [this analysis](https://github.com/rust-lang/rust/pull/71274#issuecomment-683937298) and the backlinks just above.

If someone is willing to do that for arrays, that might allow such a PR to land for arrays as well.

**Comment by RalfJung** [comment]
Yeah, crossbeam doesn't even have branches for its old versions... they might not even be able to publish fixes for them any more. I wonder if we could get them to at least yank all the old version?^^

FWIW here's what I wrote about crossbeam when we landed the aggregate check (that was in 2020):
> A dozen or more crates are broken by old versions of crossbeam-epoch using memoffset 0.2.
crossbeam-epoch 0.7.2 and 0.8 are fixed by depending on memoffset 0.5; this, too, is semver-compatible with the vast majority of users (according to download stats) but some unmaintained crates that crater tests still depend on ancient versions of crossbeam-epoch (0.3.x). Whenever I could figure out the source of the old dependency, I asked crate authors to update. There were also some panics not in crossbeam-epoch but other old crossbeam crates; I assume they, too, are fixed by newer versions (crossbeam saw tons of soundness fixes last year, so when crates depended on old versions I did not always try to pin down the exact crossbeam problem).

For the crossbeam crates causing trouble this time:
- crossbeam-queue 0.1, 0.2: if I didn't make a mistake in https://github.com/rustsec/advisory-db/blob/main/crates/crossbeam-queue/RUSTSEC-2022-0021.md, 0.2 has a semver-compatible fix. 0.2.0 is from Nov 2019.
- crossbeam-channel 0.2, 0.3, 0.4: my research for https://github.com/rustsec/advisory-db/blob/main/crates/crossbeam-channel/RUSTSEC-2022-0019.md concluded we need 0.4.3. 0.4.0 was released Nov 2019.
- crossbeam 0.2: https://github.com/rustsec/advisory-db/blob/main/crates/crossbeam/RUSTSEC-2022-0020.md says we need 0.7.x for the fix; 0.7.0 is from January 2019.
- crossbeam-deque 0.7: the 0.7 series is completely yanked. Not sure when the uninit issues were fixed. Current series is 0.8, with its initial release Oct 2020.
- crossbeam-epoch 0.8: not sure when the uninit issues were fixed. Current series is 0.9, initially released Oct 2020.

So looks like everything released since Oct 2020 has a semver-compatible fix. Possibly even more than that, depending on when exactly crossbeam-deque and crossbeam-epoch got fixed.

**Comment by RalfJung** [comment]
Another possible avenue that has been discussed is to give up on trying to *fix* old code (or getting people to fix it, by raising panics) and instead switch to mitigation mode, by reducing the amount of UB.  We could make `mem::uninitialized` synthesize a bitpattern that is actually valid at the given type (and keep the panic if the type is uninhabited). This would not fully fix UB due to `dereferenceable` assumptions for references, but it makes it much less likely that this will lead to problems.

That said, because it would not fully fix UB, I am not sure whether we should rely on this mitigation, or still also go with panics for the cases where we can get away with that.

**Comment by RalfJung** [comment]
PR for the `0x01` filling is up at https://github.com/rust-lang/rust/pull/99182.

**Comment by saethlin** [comment]
While we cannot synthesize a known-valid value for references in general, we can for slices. I think that would remove old versions of `hyper` from crater at least. And if people want to continue using those that's fine, they can upgrade at some point and get their perf back.

**Comment by RalfJung** [comment]
I think we should slightly change track here, and rather than trying to detect and root out all Rust-level UB (which at this point I consider a failed experiment), we should do what we can do mitigate the **LLVM-level UB** that can be caused by `mem::uninitialized`, while minimizing the pain for users that just followed the official advice they were given 5 years ago. https://github.com/rust-lang/rust/pull/99182 was a first step down that road, by making `mem::uninitialized` fill the returned buffer with `0x01`. Now in https://github.com/rust-lang/rust/pull/101061 I am proposing to panic in *fewer* cases than before due to the fact that LLVM UB is already mitigated.

Based on that, if https://github.com/rust-lang/rust/pull/101061 lands, my plan would be to strengthen the checks to also recurse below `Variants::Single`, and then close this issue and call it done -- to my knowledge, that mitigates all of the LLVM UB due to `aligned`, `nonnull`, or `range` annotations; there is some UB left due to `dereferenceable` annotations, but those are (to my knowledge) very unlikely to actually cause issues if the data is truly unused. In particular we do *not* have to recurse into arrays, since from what I could see, we never emit metadata for array element types.

**Comment by 5225225** [comment]
> In particular we do not have to recurse into arrays, since from what I could see, we never emit metadata for array element types.

Could we add a codegen test to *ensure* that we don't do this, leaving a comment back to this issue and saying that we must start recursing into arrays if we start emitting `nonnull` on a `[fn(); 1]` return?

It doesn't need to be done before #101061 lands, but I think it would be good to add at some point. I'll have a shot at it once I'm done with cleaning up the FCW PR.

**Comment by RalfJung** [comment]
Yes, the PR that finishes up the panics and closes this issue (the PR after https://github.com/rust-lang/rust/pull/101061) should add such a codegen test.
